File: backend/collaboration-service/internal/models/match.go
package models

import (
	"go.mongodb.org/mongo-driver/bson/primitive"
)

// Define an enum for question complexity
type QuestionComplexityEnum string

const (
	Easy   QuestionComplexityEnum = "Easy"
	Medium QuestionComplexityEnum = "Medium"
	Hard   QuestionComplexityEnum = "Hard"
)

// Question struct as defined earlier
type Question struct {
	QuestionID  string    `json:"questionId" bson:"questionId"`
	Title       string    `json:"title" bson:"title"`
	Description string    `json:"description" bson:"description"`
	Constraints string    `json:"constraints" bson:"constraints"`
	Examples    string    `json:"examples" bson:"examples"`
	Category    []string  `json:"category" bson:"category"`
	Complexity  string    `json:"complexity" bson:"complexity"`
	ImageURL    string    `json:"imageUrl" bson:"imageUrl"`
	CreatedAt   primitive.DateTime `json:"createdAt" bson:"createdAt"`
	UpdatedAt   primitive.DateTime `json:"updatedAt" bson:"updatedAt"`
}

// Match struct equivalent to the TypeScript Pair interface
type Match struct {
	ID          primitive.ObjectID     `bson:"_id,omitempty"`
	UserOne     string                  `bson:"userOne"`
	UserTwo     string                 `bson:"userTwo"`
	RoomID      string                 `bson:"room_id"`
	Complexity  []QuestionComplexityEnum `bson:"complexity"`
	Categories  []string               `bson:"categories"`
	Question    Question             `bson:"question"` 
}

File: backend/collaboration-service/models/match.ts
import { Schema, model } from "mongoose";


// Interface for Question document
export interface Question {
  questionId: string;
  title: string;
  description: string;
  constraints: string;
  examples: string;
  category: string[];
  complexity: string;
  imageUrl: string;
  createdAt: Date;
  updatedAt: Date;
}

export interface Match {
  userOne: string;
  userTwo: string;
  room_id: string;
  programming_language: string;
  complexity: string[];
  categories: string[];
  question: Question; // Single Question object
  status: string;
  createdAt: Date
}

// Mongoose schema for the Question model
export const questionSchema = new Schema<Question>({
  questionId: { type: String, required: true },
  title: { type: String, required: true },
  description: { type: String, required: false },
  constraints: { type: String, required: false },
  examples: { type: String, required: false },
  category: [{ type: String, required: false }],
  complexity: { type: String, required: false },
  imageUrl: { type: String, required: false },
  createdAt: { type: Date, default: Date.now },
  updatedAt: { type: Date, default: Date.now },
});

// Mongoose schema for the Match model, embedding a single Question
const matchSchema = new Schema<Match>({
  userOne: { type: String, required: true },
  userTwo: { type: String, required: true },
  room_id: { type: String, required: true },
  complexity: { type: [String], required: true },
  categories: { type: [String], required: true },
  programming_language: {type: String, required: true},
  question: { type: questionSchema, required: true }, // Embedded single Question document
  status: {type: String, required:true},
  createdAt: { type: Date, default: Date.now },
});

// Create models from schemas
export const QuestionModel = model<Question>("Question", questionSchema);
export const MatchModel = model<Match>("Match", matchSchema);

